core/src/main/java/org/apache/druid/math/expr/Function.java

```diff
@@ -1453,4 +1515,425 @@ public ExprEval apply(List<Expr> args, Expr.ObjectBinding bindings)
     }
   }
 
+  class StringToArrayFunction implements Function
+  {
+    @Override
+    public String name()
+    {
+      return "string_to_array";
+    }
+
+    void validateArguments(List<Expr> args)
+    {
+      if (args.size() != 2) {
+        throw new IAE("Function[%s] needs 2 argument", name());
+      }
+    }
+
+    @Override
+    public ExprEval apply(List<Expr> args, Expr.ObjectBinding bindings)
+    {
+      validateArguments(args);
+
+      final ExprEval expr = args.get(0).eval(bindings);
+      final String arrayString = expr.asString();
+      if (arrayString == null) {
+        return ExprEval.of(null);
+      }
+
+      final String split = args.get(1).eval(bindings).asString();
+      return ExprEval.ofStringArray(arrayString.split(split != null ? split : ""));
+    }
+
+    @Override
+    public Set<Expr> getScalarInputs(List<Expr> args)
+    {
+      validateArguments(args);
+      return ImmutableSet.copyOf(args);
+    }
+  }
+
+  /**
+   * Function that operates on array typed operands
+   */
+  interface ArrayFunction extends Function
+  {
+    default void validateArguments(List<Expr> args)
+    {
+      if (args.size() != 1) {
+        throw new IAE("Function[%s] needs 1 argument", name());
+      }
+    }
+
+    @Override
+    default Set<Expr> getArrayInputs(List<Expr> args)
+    {
+      validateArguments(args);
+      return ImmutableSet.of(args.get(0));
+    }
+
+    @Override
+    default Set<Expr> getScalarInputs(List<Expr> args)
+    {
+      return Collections.emptySet();
+    }
+  }
+
+  /**
+   * {@link ArraysFunction} that takes 1 array operand and 1 scalar operand
+   */
+  abstract class ArrayScalarFunction implements ArrayFunction
+  {
+    @Override
+    public void validateArguments(List<Expr> args)
+    {
+      if (args.size() != 2) {
+        throw new IAE("Function[%s] needs 2 argument", name());
+      }
+    }
+
+    @Override
+    public Set<Expr> getScalarInputs(List<Expr> args)
+    {
+      validateArguments(args);
+      return ImmutableSet.of(args.get(1));
+    }
+
+    @Override
+    public Set<Expr> getArrayInputs(List<Expr> args)
+    {
+      return ImmutableSet.of(args.get(0));
+    }
+
+    @Override
+    public ExprEval apply(List<Expr> args, Expr.ObjectBinding bindings)
+    {
+      validateArguments(args);
+      final ExprEval arrayExpr = args.get(0).eval(bindings);
+      final ExprEval scalarExpr = args.get(1).eval(bindings);
+      if (arrayExpr.asArray() == null) {
+        return ExprEval.of(null);
+      }
+      return doApply(arrayExpr, scalarExpr);
+    }
+
+    abstract ExprEval doApply(ExprEval arrayExpr, ExprEval scalarExpr);
+  }
+
+  /**
+   * {@link ArraysFunction} that takes 2 array operands
+   */
+  abstract class ArraysFunction implements ArrayFunction
+  {
+    @Override
+    public void validateArguments(List<Expr> args)
+    {
+      if (args.size() != 2) {
```

Comment:
should this and other array-handling functions check `Expr.isArray()` in `validateArguments`?

Reply:
No, I've actually removed that method from `Expr` since it wasn't actually being used. With some sort of type inference in place in the future and a more complete preservation of types of selectors that provide input bindings, we would be able to validate that the types that these expressions will evaluate to arrays, but I think just checking that the count is correct is the best we can do currently at the point this is called.